xrdpapi/xrdp-ssh-agent.c

```diff
@@ -0,0 +1,407 @@
+                    else if (errno != EINTR)
+                    {
+                        /* Error */
+                        fprintf(stderr, "Error %d on recv\n", errno);
+                        client_going = 0;
+                    }
+                }
+            }
+            else
+            {
+                /* Error */
+                fprintf(stderr, "WTSVirtualChannelRead() failed: %d\n", errno);
+                client_going = 0;
+            }
+        }
+
+        if (FD_ISSET(client_fd, &readfds))
+        {
+            /* Read from the client and write to RDP */
+            char buffer[4096];
+            ssize_t bytes_to_write = recv(client_fd, buffer, sizeof(buffer), 0);
+            if (bytes_to_write > 0)
+            {
+                char *pos = buffer;
+                while (bytes_to_write > 0)
+                {
+                    unsigned int bytes_written;
+                    int rc = WTSVirtualChannelWrite(channel,
+                                                    pos,
+                                                    bytes_to_write,
+                                                    &bytes_written);
+                    if (rc == 0)
+                    {
+                        fprintf(stderr, "WTSVirtualChannelWrite() failed: %d\n",
+                                errno);
+                        client_going = 0;
+                    }
+                    else
+                    {
+                        bytes_to_write -= bytes_written;
+                        pos += bytes_written;
+                    }
+                }
+            }
+            else if (bytes_to_write == 0)
+            {
+                /* Client has closed connection */
+                client_going = 0;
+            }
+            else
+            {
+                /* Error */
+                fprintf(stderr, "Error %d on recv\n", errno);
+                client_going = 0;
+            }
+        }
+    }
+    WTSVirtualChannelClose(channel);
+}
+
+
+int
+main(int argc, char **argv)
+{
+    /* Setup the Unix domain socket and daemon process */
+    struct sockaddr_un addr;
+    setup_ssh_agent(&addr);
+
+    /* Wait for a client to connect to the socket */
+    while (is_going)
+    {
+        fd_set readfds;
+        FD_ZERO(&readfds);
+        FD_SET(sa_uds_fd, &readfds);
+        select(FD_SETSIZE, &readfds, NULL, NULL, NULL);
+
+        /* If something connected then get it...
+         * (You can test this using "socat - UNIX-CONNECT:<udspath>".) */
+        if (FD_ISSET(sa_uds_fd, &readfds))
+        {
+            socklen_t addrsize = sizeof(addr);
+            int client_fd = accept(sa_uds_fd,
+                                   (struct sockaddr*)&addr,
+                                   &addrsize);
+            handle_connection(client_fd);
+            close(client_fd);
+        }
+    }
+
+    close(sa_uds_fd);
+    unlink(socket_name);
+
+    return 0;
+}
+
+/* vim: set sw=4:ts=4:et: */
```

Comment:
Vim mode line doesn't appear elsewhere in xrdp code but at least I don't hate the mode line. I also use Vim to edit xrdp code :)

Reply:
I've pushed a change to remove it.  I found a better way, putting the modeline in the repo's git config! https://blog.cryptomilk.org/2013/02/26/vim-modelines-in-git-config/

Reply:
thank you for letting me know. I also change the local git config!